File: backend/app/services/evaluation_engine.py
"""
Evaluation & Scoring Engine

Scores across:
- Technical depth
- Reasoning quality
- Clarity
- Confidence
- Authenticity consistency

Produces structured, unbiased output.
"""

from typing import Optional, List
from app.models.schemas import QuestionResponse, ReportResponse, SkillScore, VideoSignal


class EvaluationEngine:
    """Evaluates answers and generates final reports."""

    async def evaluate_answer(
        self,
        question: QuestionResponse,
        answer: str,
        project_context: Optional[dict] = None,
    ) -> dict:
        """
        Evaluate a single answer.
        Returns a structured evaluation with score and feedback.
        """
        # TODO: Replace with actual LLM-based evaluation
        # For MVP, use heuristic scoring

        # Basic scoring heuristics
        score = 0.5  # Default
        reasoning_depth = 0.5
        clarity = 0.5
        confidence = 0.5

        # Length-based scoring (simple heuristic)
        word_count = len(answer.split())
        if word_count > 100:
            score += 0.2
            reasoning_depth += 0.2
        elif word_count > 50:
            score += 0.1
            reasoning_depth += 0.1
        elif word_count < 20:
            score -= 0.2
            clarity -= 0.1

        # Technical keywords boost
        technical_keywords = [
            "because", "trade-off", "considered", "alternative",
            "architecture", "scalability", "performance", "security",
            "design decision", "implemented", "optimized", "tested",
        ]
        keyword_count = sum(1 for kw in technical_keywords if kw.lower() in answer.lower())
        score += min(keyword_count * 0.05, 0.2)
        reasoning_depth += min(keyword_count * 0.05, 0.2)

        # Normalize scores
        score = max(0.0, min(1.0, score))
        reasoning_depth = max(0.0, min(1.0, reasoning_depth))
        clarity = max(0.0, min(1.0, clarity))
        confidence = max(0.0, min(1.0, confidence))

        return {
            "score": score,
            "focus": question.focus,
            "reasoning_depth": reasoning_depth,
            "clarity": clarity,
            "confidence": confidence,
            "feedback": self._generate_feedback(score, question.focus),
        }

    def _generate_feedback(self, score: float, focus: str) -> str:
        """Generate feedback based on score."""
        if score >= 0.7:
            return f"Strong answer on {focus}. Good depth and reasoning."
        elif score >= 0.4:
            return f"Adequate answer on {focus}. Could provide more specific examples."
        else:
            return f"Weak answer on {focus}. Needs more depth and clarity."

    async def generate_report(
        self,
        session_id: str,
        memory: dict,
        project_context: Optional[dict] = None,
        video_signals: Optional[List[VideoSignal]] = None,
    ) -> ReportResponse:
        """Generate the final evaluation report."""
        # Calculate aggregate scores
        # In production, this would use LLM for comprehensive analysis

        skill_scores = self._calculate_skill_scores(memory)
        overall_score = self._calculate_overall_score(skill_scores)
        verdict = self._determine_verdict(overall_score)
        
        # Video behavior analysis
        behavioral_consistency = None
        if video_signals:
            behavioral_consistency = self._analyze_behavior(video_signals)

        # Project understanding (if project mode was used)
        project_understanding = None
        if project_context:
            project_understanding = self._calculate_project_understanding(memory, project_context)

        return ReportResponse(
            session_id=session_id,
            overall_score=overall_score,
            verdict=verdict,
            skill_scores=skill_scores,
            project_understanding_score=project_understanding,
            reasoning_depth_index=self._calculate_reasoning_depth(memory),
            confidence_index=self._calculate_confidence_index(memory),
            behavioral_consistency=behavioral_consistency,
            improvement_roadmap=self._generate_roadmap(memory, skill_scores),
            strengths=memory.get("strong_areas", []),
            weaknesses=memory.get("weak_areas", []),
        )

    def _calculate_skill_scores(self, memory: dict) -> List[SkillScore]:
        """Calculate scores for each skill area."""
        # Group evaluations by focus area
        focus_scores = {}
        # This is simplified - would use actual evaluation data in production
        
        default_skills = [
            SkillScore(skill="Technical Knowledge", score=0.7, feedback="Solid understanding demonstrated"),
            SkillScore(skill="Problem Solving", score=0.65, feedback="Good approach to problems"),
            SkillScore(skill="Communication", score=0.75, feedback="Clear and articulate responses"),
            SkillScore(skill="Architecture Design", score=0.6, feedback="Shows understanding of system design"),
        ]
        
        return default_skills

    def _calculate_overall_score(self, skill_scores: List[SkillScore]) -> float:
        """Calculate overall score from skill scores."""
        if not skill_scores:
            return 0.5
        return sum(s.score for s in skill_scores) / len(skill_scores)

    def _determine_verdict(self, overall_score: float) -> str:
        """Determine the final verdict."""
        if overall_score >= 0.7:
            return "Ready"
        elif overall_score >= 0.5:
            return "Borderline"
        else:
            return "Needs Work"

    def _analyze_behavior(self, signals: List[VideoSignal]) -> float:
        """Analyze behavioral consistency from video signals."""
        if not signals:
            return None
        
        # Calculate average attention and confidence
        avg_attention = sum(s.attention_score for s in signals) / len(signals)
        avg_confidence = sum(s.facial_confidence for s in signals) / len(signals)
        
        return (avg_attention + avg_confidence) / 2

    def _calculate_project_understanding(self, memory: dict, project_context: dict) -> float:
        """Calculate project understanding score."""
        # This would analyze if candidate correctly referenced project details
        return 0.7  # Placeholder

    def _calculate_reasoning_depth(self, memory: dict) -> float:
        """Calculate reasoning depth index."""
        return 0.65  # Placeholder - would aggregate from evaluations

    def _calculate_confidence_index(self, memory: dict) -> float:
        """Calculate confidence index."""
        return 0.7  # Placeholder

    def _generate_roadmap(self, memory: dict, skill_scores: List[SkillScore]) -> List[str]:
        """Generate improvement roadmap."""
        roadmap = []
        
        weak_areas = memory.get("weak_areas", [])
        for area in weak_areas:
            roadmap.append(f"Focus on improving {area} through practice and study")
        
        for skill in skill_scores:
            if skill.score < 0.6:
                roadmap.append(f"Work on {skill.skill}: {skill.feedback}")
        
        if not roadmap:
            roadmap.append("Continue building on your strong foundation")
        
        return roadmap

File: backend/src/services/evaluationEngine.js
import { AIService } from './aiService.js';
import { extractJSON } from '../utils/aiUtils.js';

const aiService = new AIService();

/**
 * EvaluationEngine - Evaluates interview answers
 * 
 * Scoring Dimensions:
 * - Correctness (technical accuracy)
 * - Depth (level of understanding)
 * - Clarity (communication quality)
 * - Practical Understanding (real-world application)
 * - Confidence (certainty in answer)
 */
export class EvaluationEngine {
    /**
     * Evaluate a single answer
     */
    async evaluateAnswer({ question, answer, context, type }) {
        const systemPrompt = this.buildEvaluationPrompt(type);

        const userPrompt = `
Question: "${question.text}"
Expected Points: ${question.expectedPoints?.join(', ') || 'N/A'}
Topic: ${question.topic}
Difficulty: ${question.difficulty}

Candidate's Answer:
"${answer}"

Context:
- Role: ${context?.role || 'Software Engineer'}
- Experience: ${context?.experience || 'mid'}

Evaluate this answer comprehensively.
`;

        try {
            const response = await aiService.generate({
                systemPrompt,
                userPrompt,
                responseFormat: 'json'
            });

            const evaluation = extractJSON(response);
            if (!evaluation) throw new Error('Failed to parse AI evaluation response');


            return {
                questionId: question.id,
                scores: {
                    correctness: evaluation.correctness || 5,
                    depth: evaluation.depth || 5,
                    clarity: evaluation.clarity || 5,
                    practicalUnderstanding: evaluation.practicalUnderstanding || 5,
                    confidence: evaluation.confidence || 5
                },
                overallScore: evaluation.overallScore || this.calculateOverallScore(evaluation),
                feedback: evaluation.feedback || '',
                strongPoints: evaluation.strongPoints || [],
                weakPoints: evaluation.weakPoints || [],
                missedConcepts: evaluation.missedConcepts || [],
                followUpSuggestion: evaluation.followUpSuggestion || null
            };
        } catch (error) {
            console.error('Evaluation failed:', error);
            return this.getDefaultEvaluation(question.id);
        }
    }

    /**
     * Evaluate an entire interview session
     */
    async evaluateSession({ sessionId, questions, answers, context }) {
        // Evaluate each answer
        const evaluations = [];

        for (let i = 0; i < questions.length; i++) {
            const question = questions[i];
            const answer = answers[i]?.answer || '';

            const evaluation = await this.evaluateAnswer({
                question,
                answer,
                context,
                type: question.type
            });

            evaluations.push(evaluation);
        }

        // Aggregate results
        const aggregated = this.aggregateEvaluations(evaluations);

        // Generate overall assessment
        const overallAssessment = await this.generateOverallAssessment({
            evaluations,
            context,
            questions
        });

        return {
            sessionId,
            evaluations,
            ...aggregated,
            ...overallAssessment,
            completedAt: new Date().toISOString()
        };
    }

    /**
     * Build evaluation prompt based on interview type
     */
    buildEvaluationPrompt(type) {
        const base = `You are an expert interviewer evaluating candidate responses.

Score each dimension from 0-10:
- correctness: Technical accuracy of the answer
- depth: Level of understanding demonstrated
- clarity: How clearly the answer was communicated
- practicalUnderstanding: Real-world application knowledge
- confidence: How confident the candidate seemed

Respond in JSON format:
{
  "correctness": 0-10,
  "depth": 0-10,
  "clarity": 0-10,
  "practicalUnderstanding": 0-10,
  "confidence": 0-10,
  "overallScore": 0-10,
  "feedback": "Constructive feedback for the candidate",
  "strongPoints": ["What they did well"],
  "weakPoints": ["Areas for improvement"],
  "missedConcepts": ["Important concepts not mentioned"],
  "followUpSuggestion": "A good follow-up question to ask"
}`;

        const typeAdditions = {
            technical: '\n\nFocus on technical accuracy, problem-solving approach, and code quality understanding.',
            hr: '\n\nFocus on communication skills, emotional intelligence, and behavioral indicators.',
            project: '\n\nFocus on architectural decisions, practical experience, and problem-solving during development.'
        };

        return base + (typeAdditions[type] || typeAdditions.technical);
    }

    /**
     * Calculate overall score from dimensions
     */
    calculateOverallScore(evaluation) {
        const weights = {
            correctness: 0.3,
            depth: 0.25,
            clarity: 0.2,
            practicalUnderstanding: 0.15,
            confidence: 0.1
        };

        let score = 0;
        for (const [key, weight] of Object.entries(weights)) {
            score += (evaluation[key] || 5) * weight;
        }

        return Math.round(score * 10) / 10;
    }

    /**
     * Aggregate multiple evaluations
     */
    aggregateEvaluations(evaluations) {
        if (evaluations.length === 0) {
            return { averageScore: 0, skillBreakdown: {} };
        }

        // Calculate averages
        const totals = {
            correctness: 0,
            depth: 0,
            clarity: 0,
            practicalUnderstanding: 0,
            confidence: 0,
            overall: 0
        };

        const allStrongPoints = [];
        const allWeakPoints = [];
        const allMissedConcepts = [];

        evaluations.forEach(evaluation => {
            totals.correctness += evaluation.scores.correctness;
            totals.depth += evaluation.scores.depth;
            totals.clarity += evaluation.scores.clarity;
            totals.practicalUnderstanding += evaluation.scores.practicalUnderstanding;
            totals.confidence += evaluation.scores.confidence;
            totals.overall += evaluation.overallScore;

            allStrongPoints.push(...evaluation.strongPoints);
            allWeakPoints.push(...evaluation.weakPoints);
            allMissedConcepts.push(...evaluation.missedConcepts);
        });

        const count = evaluations.length;

        return {
            averageScore: Math.round((totals.overall / count) * 10),
            skillBreakdown: {
                correctness: Math.round((totals.correctness / count) * 10),
                depth: Math.round((totals.depth / count) * 10),
                clarity: Math.round((totals.clarity / count) * 10),
                practicalUnderstanding: Math.round((totals.practicalUnderstanding / count) * 10),
                confidence: Math.round((totals.confidence / count) * 10)
            },
            strongAreas: [...new Set(allStrongPoints)].slice(0, 5),
            weakAreas: [...new Set(allWeakPoints)].slice(0, 5),
            missedConcepts: [...new Set(allMissedConcepts)].slice(0, 5)
        };
    }

    /**
     * Generate overall assessment
     */
    async generateOverallAssessment({ evaluations, context, questions }) {
        const averageScore = evaluations.reduce((sum, e) => sum + e.overallScore, 0) / evaluations.length;

        let recommendation = '';
        if (averageScore >= 8) {
            recommendation = 'Strong candidate. Ready for the next round.';
        } else if (averageScore >= 6) {
            recommendation = 'Good potential. Some areas need improvement.';
        } else if (averageScore >= 4) {
            recommendation = 'Needs more preparation. Review weak areas.';
        } else {
            recommendation = 'Significant gaps. Recommend focused practice.';
        }

        return {
            overallAssessment: {
                score: Math.round(averageScore * 10),
                recommendation,
                readinessLevel: averageScore >= 7 ? 'high' : averageScore >= 5 ? 'medium' : 'low'
            }
        };
    }

    /**
     * Evaluate communication quality
     */
    async evaluateCommunication(answers) {
        // Analyze communication patterns across all answers
        const metrics = {
            clarity: 0,
            structure: 0,
            conciseness: 0,
            vocabulary: 0
        };

        answers.forEach(answer => {
            const text = answer.answer || '';

            // Simple heuristics (in production, use AI)
            metrics.structure += text.includes('\n') || text.length > 100 ? 7 : 5;
            metrics.clarity += text.split('.').length > 2 ? 7 : 5;
            metrics.conciseness += text.length < 500 ? 7 : text.length < 1000 ? 5 : 3;
            metrics.vocabulary += 7; // Placeholder
        });

        const count = answers.length || 1;

        return {
            clarity: Math.round(metrics.clarity / count),
            structure: Math.round(metrics.structure / count),
            conciseness: Math.round(metrics.conciseness / count),
            vocabulary: Math.round(metrics.vocabulary / count),
            overall: Math.round(Object.values(metrics).reduce((a, b) => a + b) / (4 * count))
        };
    }

    /**
     * Get default evaluation if AI fails
     */
    getDefaultEvaluation(questionId) {
        return {
            questionId,
            scores: {
                correctness: 5,
                depth: 5,
                clarity: 5,
                practicalUnderstanding: 5,
                confidence: 5
            },
            overallScore: 5,
            feedback: 'Unable to evaluate automatically. Please review manually.',
            strongPoints: [],
            weakPoints: [],
            missedConcepts: [],
            followUpSuggestion: null
        };
    }
}
